File: Java/103.二叉树的锯齿形层次遍历.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        if(root == null) return res;
        Queue<TreeNode> qu = new LinkedList<>();
        //标记正序、反序层
        boolean flag = true;
        qu.offer(root);
        int cnt = 0;
        while(!qu.isEmpty()){
            cnt = qu.size();
            List<Integer> level = new ArrayList<>();
            for(int i = 0; i < cnt; i++){
                TreeNode temp = qu.poll();
                if(temp.left != null) qu.offer(temp.left);
                if(temp.right != null) qu.offer(temp.right);
                if(flag) level.add(temp.val);
                else level.add(0,temp.val);
            }
            res.add(level);
            flag = !flag;
        }
        return res;
    }
}

File: CPP/103. 二叉树的锯齿形层次遍历.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
#include <queue>
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> res;
        if(root == NULL) return res;
        queue<TreeNode*> qu;
        qu.push(root);
        bool order = false;
        while(!qu.empty()){
            int n = qu.size();
            vector<int> temp;
            for(int i = 0; i < n; i++){
                TreeNode* current = qu.front();
                qu.pop();
                if(current->left != NULL) qu.push(current->left);
                if(current->right != NULL)qu.push(current->right);
                if(order){
                    temp.insert(temp.begin(),1,current->val);
                }else{
                    temp.push_back(current->val);
                }
            }
            order = !order;
            res.push_back(temp);
        }
        return res;        
    }
};